### PR TITLE
UnstructuredGrid volume rendering

### LINKED ISSUE #3801: Volume Rendering for Unstructured Grid
### Describe the feature you would like to be added.

Volume rendering for unstructured grid is needed. Upstream in VTK, it has this `vtkUnstructuredGridVolumeMapper`  which can fullfill this work. Please add this in pyvista.

### Links to VTK Documentation, Examples, or Class Definitions.

Here is a [link](https://vtk.org/doc/release/6.2/html/classvtkUnstructuredGridVolumeMapper.html) for this class in VTK.

### Pseudocode or Screenshots

```python
Insert code here
```
Or drag your screenshot here!

### PR DESCRIPTION
**Pull request by akaszynski**
Resolve #3801 by adding volume rendering for unstructured grids.

```py
import pyvista
pyvista.set_plot_theme('document')
from pyvista import examples
mesh = examples.download_letter_a()
mesh['scalars'] = mesh.points[:, 1]

pl = pyvista.Plotter()
pl.add_volume(mesh, opacity_unit_distance=0.1, cmap='plasma')
pl.show()
```

https://user-images.githubusercontent.com/11981631/215971577-1eabd3aa-b07e-49d0-9375-f47a0e3de046.mp4

### CODE REVIEW

**Review by banesullivan — CHANGES_REQUESTED** [review]
I made some minor changes to docstrings and forcibly overide the mapper choice if a UGrid is used.

Further, I often see this VTK err:
```
2023-02-01 09:29:39.344 (   1.233s) [        AAA3D280]vtkUnstructuredGridBuny:930   WARN| vtkUnstructuredGridBunykRayCastFunction (0x557064a77c40): Input contains more than tetrahedra - only tetrahedra are supported
```

a `tesselate()` filter does the trick, but can we add this same tetrahedral check to PyVista and raise an actual error to prevent a failed VTK pipeline?

I would actually really like to land this for 0.38 today if you feel confident in this, @akaszynski

**Comment by banesullivan** [comment]
I looked into how that warning is generated, and it may be tough to implement Python side without a serious performance hit.

**Comment by banesullivan** [comment]
@akaszynski, we could just run a `triangulate()` or `tessellate()` filter automatically for any UnstructuredGrid input... I think this is what ParaView does to make volume rendering of any data type possible.

**Comment by banesullivan** [comment]
Okay, done. This know allows `add_mesh()` to take pretty much any 3D data set and volume render it. There is a performance cost to unstructured grid volume rendering, but that is explicitly warned in the documentation.

IMO, this is much better and less confusing than only supporting volume rendering for a handful of data types.

@akaszynski, please review my changes. I'd love to land this asap for 0.38 as a ton of people have requested this

File: pyvista/plotting/plotting.py [hunk 1]
```diff
@@ -3867,16 +3891,36 @@ def add_volume(
...
+                f'Type {type(volume)} not supported for volume rendering as it is not 3D.'
             )
+        elif not isinstance(
+            volume, (pyvista.UniformGrid, pyvista.RectilinearGrid, pyvista.UnstructuredGrid)
+        ):
+            volume = volume.cast_to_unstructured_grid()
+
+        # Override mapper choice for UnstructuredGrid
+        if isinstance(volume, pyvista.UnstructuredGrid):
+            volume = volume.triangulate()  # forcibly triangulate, should be inexpensive
```
Comment (banesullivan) [inline pyvista/plotting/plotting.py:3911]: I'm pretty sure VTK and ParaView use this as it's fairly performant.

The tessellate() filter is expensive. Users can opt-in to that on their own

**Review by banesullivan — DISMISSED** [review]

File: pyvista/plotting/plotting.py (hunk 1 above)
Comment (akaszynski) [reply pyvista/plotting/plotting.py:3911]: Faster to check and then `triangulate`:

```py
import pyvista as pv


ugrid = pv.UniformGrid(dimensions=(50, 50, 50)).cast_to_unstructured_grid()
ugrid_tet = ugrid.triangulate()


def always_triangulate(ugrid):
    return ugrid.triangulate()


def check_then_triangulate(ugrid):
    if not (ugrid.celltypes == pv.celltype.CellType.TETRA).all():
        return ugrid.triangulate()
    return ugrid

"""
>>> timeit check_then_triangulate(ugrid_tet)
63.7 µs ± 74.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

timeit always_triangulate(ugrid_tet)
485 µs ± 4.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> timeit check_then_triangulate(ugrid)
2.21 s ± 34.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

timeit always_triangulate(ugrid)
2.23 s ± 7.96 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

"""
```

Comment (banesullivan) [reply pyvista/plotting/plotting.py:3911]: Good by me, can you add that right quick?

Comment (akaszynski) [reply pyvista/plotting/plotting.py:3911]:
```suggestion
            # Unstructured grid must be all tetrahedrals
            if not (volume.celltypes == pv.celltype.CellType.TETRA).all():
                return volume.triangulate()
```

**Review by banesullivan — APPROVED** [review]

**Pull request merged**